src/interchange/avro.rs

```diff
@@ -61,46 +61,82 @@ pub fn validate_key_schema(key_schema: &str, value_desc: &RelationDesc) -> Resul
-        // have the same schema.
-        let row_schema = match node.inner {
-            SchemaPiece::Record { fields, .. } => {
-                let before = fields.iter().find(|f| f.name == "before");
-                let after = fields.iter().find(|f| f.name == "after");
-                match (before, after) {
-                    (Some(before), Some(after)) => {
-                        let left = node.step(&before.schema);
-                        let right = node.step(&after.schema);
-                        if let Some((left, right)) = first_mismatched_schema_types(left, right) {
-                            bail!(
-                            "source schema has mismatched 'before' and 'after' schemas: before={:?} after={:?}",
-                            left.inner,
-                            right.inner
-                        )
+    let row_schema = match envelope {
+        EnvelopeType::Debezium => {
+            // The top-level record needs to be a diff "envelope" that contains
+            // `before` and `after` fields, where the `before` and `after` fields
+            // have the same schema.
+            let row_schema = match node.inner {
+                SchemaPiece::Record { fields, .. } => {
+                    let before = fields.iter().find(|f| f.name == "before");
+                    let after = fields.iter().find(|f| f.name == "after");
+                    match (before, after) {
+                        (Some(before), Some(after)) => {
+                            let left = node.step(&before.schema);
+                            let right = node.step(&after.schema);
+                            if let Some((left, right)) = first_mismatched_schema_types(left, right)
+                            {
+                                bail!(
+                                "source schema has mismatched 'before' and 'after' schemas: before={:?} after={:?}",
+                                left.inner,
+                                right.inner
+                            )
+                            }
+                            &before.schema
+                        }
+                        (None, _) => bail!("source schema is missing 'before' field"),
+                        (_, None) => bail!("source schema is missing 'after' field"),
+                    }
+                }
+                _ => bail!("source schema does not match required envelope format"),
+            };
+            // The "row" schema used by the `before` and `after` fields needs to be
+            // a nullable record type.
+            match row_schema.get_piece_and_name(&schema).0 {
+                SchemaPiece::Union(us) => {
+                    if us.variants().len() != 2 {
+                        bail!("source schema 'before'/'after' fields are not of expected type");
+                    }
+                    let has_null = us.variants().iter().any(|s| is_null(s));
+                    let record =
```

Comment:
(This comment applies to the similar code in `EnvelopeType::Upsert` below, but GitHub won't let me add it there because its diffing algorithm got confused.)

Do you know what the name of the variant should be?

This is fine for now, but I will add a way to look up union schemas by name in the Avro library and do things that way. Matching on record type is a horrible hack that we were doing for Debezium because we didn't properly support named types in Avro schemas; I will change it in both places after you land this.

Reply:
There is no fixed name for the upsert variant schema because arguably "envelope" is a misnomer, and it just refers to the kafka convention if you ingest records with a key and a value, later rows with the same key will be treated as an update or deletion of earlier data. Basically, the value schema is either an arbitrary user-defined record or a union of null + an arbitrary user-defined record.

You can try out the key-value kafka topics using instructions here. https://docs.confluent.io/3.0.0/schema-registry/docs/serializer-formatter.html#formatter